### PR TITLE
Issue 509 & 313

### LINKED ISSUE #313: Redesign of the logic for tumor only and germline samples
The current approach is to perform variant calling (germline mode) for the samples even
if they are paired (tumor-normal). So, for instance if we have

SAMPLE1 0
SAMPLE2 1 
SAMPLE3 1

Sarek will perform variant calling in somatic mode for SAMPLE2_vs_SAMPLE1 and in germline mode
for SAMPLE1, SAMPLE2 and SAMPLE3. I believe it is not necessary to perform variant calling
in germline mode for SAMPLE1 and SAMPLE2 (specially for SAMPLE1). By changing this behaviour
we would save a considerable amount of computational time and resources (specially for big datasets). 

This, of course, may vary depending on the tools used. Any thoughts on this? Technically, it would not
require a lot of changes in the code to change the current behaviour.

### LINKED ISSUE #509: Select only somatic/germline[FEATURE]
In order to speed up the (wonderful!) _Sarek_ pipeline, could be it considered to add an option to process somatic *OR* germline variants and not all indistinctly?

Maybe something like `--step somatic/germline`?

Thanks

### PR DESCRIPTION
**Pull request by FriederikeHanssen**
Closes #509 
closes #313 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: workflows/sarek.nf [hunk 1]
```diff
@@ -628,6 +627,22 @@ workflow SAREK {
         // and remove patient ID field & null value for further processing [meta1, [cram1,crai1]] [meta2, [cram2,crai2]]
         cram_variant_calling_tumor_only = cram_variant_calling_tumor_filtered.transpose().map{ it -> [it[1], it[2], it[3]] }
 
+        if(params.skip_germline_for_somatic){
```
Comment (FriederikeHanssen) [inline workflows/sarek.nf:630]: not too sure about the parameter name, I was also considering `skip_germline_for_paired` or something along those lines

Comment (maxulysse) [reply workflows/sarek.nf:630]: `only_paired_variant_calling`?

Comment (FriederikeHanssen) [reply workflows/sarek.nf:630]: nice, this is really says what it does

File: CHANGELOG.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - [#507](https://github.com/nf-core/sarek/pull/507) - Subway map for building indexes
 - [#512](https://github.com/nf-core/sarek/pull/512) - Subway map for pipeline
 - [#522](https://github.com/nf-core/sarek/pull/522) - Add QC for vcf files & MultiQC
+- [#533](https://github.com/nf-core/sarek/pull/533) - Add param `--skip_germline_for_somatic` to allow skipping of germline variantcalling for paired samples
```
Comment (maxulysse) [inline CHANGELOG.md:24]:
```suggestion
- [#533](https://github.com/nf-core/sarek/pull/533) - Add param `--only_paired_variant_calling` to allow skipping of germline variantcalling for paired samples
```

**Review by maxulysse — APPROVED** [review]
Easy to do indeed.
Should we have some tests for that?

**Comment by FriederikeHanssen** [comment]
> Easy to do indeed. Should we have some tests for that?

Probably a good idea. Let me add some

File: tests/csv/3.0/variantcalling_channels.csv [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+test1,XX,1,sample2,https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test2.paired_end.recalibrated.sorted.cram,https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test2.paired_end.recalibrated.sorted.cram.crai
+test3,XX,0,sample3,https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test.paired_end.recalibrated.sorted.cram,https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test.paired_end.recalibrated.sorted.cram.crai
+test3,XX,1,sample4,https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test2.paired_end.recalibrated.sorted.cram,https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test2.paired_end.recalibrated.sorted.cram.crai
```
Comment (maxulysse) [inline tests/csv/3.0/variantcalling_channels.csv:5]: why 2 pairs?

File: tests/test_only_paired_VC.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    - path: results/variant_calling/sample3/strelka/sample3.genome.vcf.gz
+    - path: results/variant_calling/sample3/strelka/sample3.genome.vcf.gz.tbi
+    - path: results/variant_calling/sample2/strelka/sample2.variants.vcf.gz
+    - path: results/variant_calling/sample2/strelka/sample2.variants.vcf.gz.tbi
+    - path: results/variant_calling/sample2/strelka/sample2.genome.vcf.gz
+    - path: results/variant_calling/sample2/strelka/sample2.genome.vcf.gz.tbi
+    - path: results/variant_calling/sample4_vs_sample3/strelka/sample4_vs_sample3.somatic_indels.vcf.gz
+    - path: results/variant_calling/sample4_vs_sample3/strelka/sample4_vs_sample3.somatic_indels.vcf.gz.tbi
+    - path: results/variant_calling/sample4_vs_sample3/strelka/sample4_vs_sample3.somatic_snvs.vcf.gz
+    - path: results/variant_calling/sample4_vs_sample3/strelka/sample4_vs_sample3.somatic_snvs.vcf.gz.tbi
```
Comment (maxulysse) [inline tests/test_only_paired_VC.yml:45]: aren't we missing sample1?

Comment (FriederikeHanssen) [reply tests/test_only_paired_VC.yml:45]: yes

File: tests/csv/3.0/variantcalling_channels.csv (hunk 3 above)
Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: I used to make sure that sample3 is not run. I think I am missing a couple of check sthough

Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: So basically I justed wanted to make sure we can test all options of the channel magic, that is why two "pairs". Single-normal, single-tumor, and a paired one. Or do you think that is too extensive?

Comment (maxulysse) [reply tests/csv/3.0/variantcalling_channels.csv:5]: Ooooh I see.
May I suggest an alternative solution, not sure if it'll work or not, but could be worth a try
Have 3 different samplesheets:
- single normal
- single tumor
- pair
That way we can use those in different tests.
but call it here in this case with `--input samplesheet1.csv,samplesheet2.csv,samplesheet3.csv`
I know it worked on local with previous sarek version, might be working here as well in this case

Comment (maxulysse) [reply tests/csv/3.0/variantcalling_channels.csv:5]: otherwise, happy with this idea, that's good and smart

Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: ah nice, I forgot again about "merging" of input csvs. I will try it then we can just use the ones we have

Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: Don't think that works anymore: ```No such file: /Users/monarchy/Projects/Coding/sarek/tests/csv/3.0/recalibrated_germline.csv,tests/csv/3.0/recalibrated_tumoronly.csv,tests/csv/3.0/recalibrated_somatic.csv``` 

with providing it with:
```nextflow run main.nf -profile test,variantcalling_channels,docker --input tests/csv/3.0/recalibrated_germline.csv,tests/csv/3.0/recalibrated_tumoronly.csv,tests/csv/3.0/recalibrated_somatic.csv```

Is that a feature you need again?

Comment (maxulysse) [reply tests/csv/3.0/variantcalling_channels.csv:5]: quotes?
`nextflow run main.nf -profile test,variantcalling_channels,docker --input "tests/csv/3.0/recalibrated_germline.csv,tests/csv/3.0/recalibrated_tumoronly.csv,tests/csv/3.0/recalibrated_somatic.csv"`

Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: nope neither single nor double quotes. Trying to find the place in sakre 2.7.1 that implementted this

Comment (maxulysse) [reply tests/csv/3.0/variantcalling_channels.csv:5]: you can stop looking, managed to run:
```
nextflow run . -profile test,docker --input "tests/csv/3.0/fastq_*.csv"
```
There's no `,` involved, just `*` so won't work on https path, just s3 + local

Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: hm, ok what works locally now is ```nextflow run main.nf -profile test,variantcalling_channels,docker --input tests/csv/3.0/recalibrated_{germline,tumoronly,somatic}.csv --tools strelka ```

Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: it think it is the `if (params.input) csv_file = file(params.input)` line and how this is expanded https://www.nextflow.io/docs/latest/script.html#files-and-i-o

Comment (maxulysse) [reply tests/csv/3.0/variantcalling_channels.csv:5]: yes, we can use globs, so `*` or `{}`, I'd say probably `?`, but not https definitively

Comment (FriederikeHanssen) [reply tests/csv/3.0/variantcalling_channels.csv:5]: ah never mind actually doesn't run the processes except for the first file.

**Comment by FriederikeHanssen** [comment]
ok one last thing, there is actually a `recalibrated.csv` that has a bunch of file and almost fulfills all constraints. I would tweak that one for this use case ( we are not using it so far for CI tests)

**Review by maxulysse — APPROVED** [review]
pure beauty

**Pull request merged**